File: src/main/python/AoC2024_02.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 2
#

import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Input = list[list[int]]
Output1 = int
Output2 = int


TEST = """\
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [list(map(int, line.split())) for line in input_data]

    def safe(self, levels: list[int]) -> bool:
        diffs = [levels[i + 1] - levels[i] for i in range(len(levels) - 1)]
        return all(1 <= diff <= 3 for diff in diffs) or all(
            -1 >= diff >= -3 for diff in diffs
        )

    def part_1(self, reports: Input) -> Output1:
        return sum(self.safe(report) for report in reports)

    def part_2(self, reports: Input) -> Output2:
        return sum(
            any(
                self.safe(report[:i] + report[i + 1 :])
                for i in range(len(report))
            )
            for report in reports
        )

    @aoc_samples(
        (
            ("part_1", TEST, 2),
            ("part_2", TEST, 4),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 2)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_02.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;

import com.github.pareronia.aoc.itertools.IterTools;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class AoC2024_02
        extends SolutionBase<List<List<Integer>>, Integer, Integer> {
    
    private AoC2024_02(final boolean debug) {
        super(debug);
    }
    
    public static AoC2024_02 create() {
        return new AoC2024_02(false);
    }
    
    public static AoC2024_02 createDebug() {
        return new AoC2024_02(true);
    }
    
    @Override
    protected List<List<Integer>> parseInput(final List<String> inputs) {
        return inputs.stream()
            .map(line -> Arrays.stream(line.split(" "))
                    .map(Integer::parseInt)
                    .toList())
            .toList();
    }
    
    private boolean safe(final List<Integer> levels) {
        final List<Integer> diffs = IterTools.pairwise(levels.iterator()).stream()
            .map(w -> w.second() - w.first())
            .toList();
        return diffs.stream().allMatch(diff -> 1 <= diff && diff <= 3)
            || diffs.stream().allMatch(diff -> -1 >= diff && diff >= -3);
    }
    
    @Override
    public Integer solvePart1(final List<List<Integer>> reports) {
        return (int) reports.stream().filter(this::safe).count();
    }
    
    @Override
    public Integer solvePart2(final List<List<Integer>> reports) {
        return (int) reports.stream()
            .filter(report ->
                range(report.size()).intStream()
                    .mapToObj(i -> {
                        final List<Integer> tmp = new ArrayList<>(report);
                        tmp.remove(i);
                        return tmp;
                    })
                    .anyMatch(this::safe)
            ).count();
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "2"),
        @Sample(method = "part2", input = TEST, expected = "4"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2024_02.create().run();
    }

    private static final String TEST = """
            7 6 4 2 1
            1 2 7 8 9
            9 7 6 2 1
            1 3 2 4 5
            8 6 4 4 1
            1 3 6 7 9
            """;
}